### DISCUSSION TITLE
[CLI] tootctl commands fail with postgres FATAL error

### DISCUSSION THREAD

**Discussion by LanceMcCarthy** [mastodon/mastodon#22890] [General]
### Steps to reproduce the problem

1. `cd` into the bin directory
2. Run `./tootctl media remove --days=7`
3. Observe error message `FATAL: Peer authentication failed for user "mastodon"`

I can also replicate with any CLI command, here's an example trying to get the version

```bash
./tootctl --version
...[removed stacktrace]...
/var/lib/gems/2.7.0/gems/pg-1.3.5/lib/pg/connection.rb:637:in `async_connect_or_reset': connection to server on socket "/var/run/postgresql/.s.PGSQL.5432" failed: FATAL:  Peer authentication failed for user "mastodon" (PG::ConnectionBad)
```

### Expected behaviour

The `tootctl` command executes without error and delete media

### Actual behavior

There is an error about the postgres user `mastodon` not being able to authenticate with postgres using `peer` authentication method

### Detailed description

In researching this, [I saw some options](https://stackoverflow.com/questions/18664074/getting-error-peer-authentication-failed-for-user-postgres-when-trying-to-ge) to change `peer` to `password/md5`, but I dont want to take any of these drastic options until I understand more about why it is happening.

I can confirm that the  pg_hba.conf file has 127.0.0.0/24 in the accepted hosts under the IPv4 section, so it's not a problem with network access. This just seems ot be a problem with the mastodon user attempting to run the delete commands that `tootctl media remove` needs

## Solution

Edit `/etc/postgresql/[version]/main/pg_hba.conf` and change `peer`  to `scram-sha-256` for the local connection. See my reply below with screenshots https://github.com/mastodon/mastodon/issues/22811#issuecomment-1367526230

### Specifications

Latest stable mastodon (v3.5)
Running on the metal (not a container)
Ubuntu 20.04

**Comment by cadamini** [comment]
Peer is only supported on local connections. (from the [psql docs](https://www.postgresql.org/docs/10/auth-methods.html#AUTH-PEER)). The problem might be your pg_hba.conf config file using peer instead of md5. You linked the related SO post. Not sure if changing this is a bad thing, see https://stackoverflow.com/questions/17443379/psql-fatal-peer-authentication-failed-for-user-dev

**Comment by LanceMcCarthy** [comment]
I can confirm it is using peer, but if I change it to md5 is has to know where to get the hashed password that used when the mastodon user was first created, correct?  Or is that already stored in a database setting that postgress knows to get the md5 string from?

**Comment by cadamini** [comment]
If I am not totally wrong, the user password is set on user creation. You need to set an environment variable that is read according to the database.yml file, see https://github.com/mastodon/mastodon/blob/ef4d29c8791086b11f6e36aa121ff5c9b5fa0103/config/database.yml#L32

**Comment by LanceMcCarthy** [comment]
> If I am not totally wrong, the user password is set on user creation. You need to set an environment variable that is read according to the database.yml file,

Yes, that env was where the DB user and DB passwords were set.  The reason I supposed `peer` should work is because both the mastodon server (and sidekiq jobs) are all on localhost using the same vNIC

So according to that db yaml config, I'm not sure if I need to use `md5` or `password`... but it's a 50%/50% chance :D I'll give it a shot first thing in the morning. Thanks for the rubber duck debugging!

**Comment by LanceMcCarthy** [comment]
No luck, after the change.

Here's my starting pg_hba.conf content

```
# TYPE  DATABASE    USER   ADDRESS    METHOD

# "local" is for Unix domain socket connections only
local   all             all                                     peer

# IPv4 local connections:
host    all  all  127.0.0.1/32                      scram-sha-256
# IPv6 local connections:
host    all  all    ::1/128                             scram-sha-256
# Allow replication connections from localhost, by a user with the replication privilege.
local   replication     all                            peer
host    replication     all    127.0.0.1/32    scram-sha-256
host    replication     all    ::1/128             scram-sha-256
```

I changed that first line from peer to md5 and ran the following command again

`$ RAILS_ENV=production /var/www/mastodon/bin/tootctl media remove --days=7`

here's the error (it looks the same as before)

```
warning: parser/current is loading parser/ruby27, which recognizes2.7.6-compliant syntax, but you are running 2.7.0.
Please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
/var/lib/gems/2.7.0/gems/activerecord-6.1.6/lib/active_record/connection_adapters/postgresql_adapter.rb:78: warning: Using the last argument as keyword parameters is 
deprecated; maybe ** should be added to the call
/var/lib/gems/2.7.0/gems/pg-1.3.5/lib/pg.rb:68: warning: The called method `connect' is defined here
...redacted trace...
/var/lib/gems/2.7.0/gems/pg-1.3.5/lib/pg/connection.rb:637:in `async_connect_or_reset': connection to server on socket "/var/run/postgresql/.s.PGSQL.5432" failed: FATAL:  Peer authentication failed for user "mastodon" (PG::ConnectionBad)
...redacted trace...
/var/lib/gems/2.7.0/gems/activerecord-6.1.6/lib/active_record/connection_adapters/postgresql_adapter.rb:83:in `rescue in new_client': connection to server on socket "
/var/run/postgresql/.s.PGSQL.5432" failed: FATAL:  Peer authentication failed for user "mastodon" (ActiveRecord::ConnectionNotEstablished)
```

**Comment by LanceMcCarthy** [comment]
FIXED IT 🔥
Needed to switch from `peer` to `scram-sha-256` for the method.

![image](https://user-images.githubusercontent.com/3520532/209997280-1c9b3e01-1cc4-493e-88d3-01fa7c77cc2f.png)


Side Note: I also had to use `sudo` for the media command. I'm not sure if this was required before, or if the CLI docs assume you're running them as **su**. least my cloud storage isnt going to keep

![image](https://user-images.githubusercontent.com/3520532/209997918-57137912-bdbb-4319-8ad1-2dbb03672e92.png)

In any case, I'm finally happy that my cloud costs for storage wont keep exploding because I can't delete old media :D
